File: day3.py
from beartype import beartype

from util import load_input


instructions = load_input(3)


@beartype
def deliveries(instructions: str) -> set[int]:
    """Find the number of houses that got at least one present.

    >>> num_deliveries(">")
    2

    >>> num_deliveries("^>v<")
    4

    >>> num_deliveries("^v^v^v^v^v")
    2
    """
    visited = set([(0, 0)])
    curr_pos = (0, 0)

    for ins in instructions:
        cx, cy = curr_pos
        match ins:
            case "^":
                curr_pos = (cx, cy - 1)
            case "v":
                curr_pos = (cx, cy + 1)
            case ">":
                curr_pos = (cx + 1, cy)
            case "<":
                curr_pos = (cx - 1, cy)
            case _:
                raise ValueError(f"Unknown instruction {ins}")
        visited.add(curr_pos)

    return visited

print(f"Part 1: {len(deliveries(instructions))}")

normal = ""
robo = ""

for i, ins in enumerate(instructions):
    if i % 2 == 0:
        normal += ins
    else:
        robo += ins

houses = deliveries(normal) | deliveries(robo)
print(f"Part 2: {len(houses)}")

File: src/day3.jl
include("./util.jl")
using Test: @test

instructions = util.load_input(3)

function deliveries(instructions::String)::Set{Tuple{Int,Int}}
    visited = Set()
    curr_pos = (0, 0)
    push!(visited, curr_pos)

    for ins in instructions
        cx, cy = curr_pos
        if ins == '^'
            curr_pos = (cx, cy - 1)
        elseif ins == 'v'
            curr_pos = (cx, cy + 1)
        elseif ins == '>'
            curr_pos = (cx + 1, cy)
        elseif ins == '<'
            curr_pos = (cx - 1, cy)
        else
            throw(DomainError("Unknown instruction $ins"))
        end

        push!(visited, curr_pos)
    end
    visited
end

num_deliveries(instructions) = length(deliveries(instructions))

@test num_deliveries(">") == 2
@test num_deliveries("^>v<") == 4
@test num_deliveries("^v^v^v^v^v") == 2

println("Part 1: $(num_deliveries(instructions))")

normal = ""
robo = ""

for (i, ins) in enumerate(instructions)
    if i % 2 == 0
        global normal *= ins
    else
        global robo *= ins
    end
end

houses = union(deliveries(normal), deliveries(robo))
println("Part 2: $(length(houses))")